flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -377,6 +377,10 @@ private void restartTasks(
         final Set<ExecutionVertexID> verticesToRestart =
                 executionVertexVersioner.getUnmodifiedExecutionVertices(executionVertexVersions);
 
+        if (verticesToRestart.isEmpty()) {
+            return;
```

Comment:
I'm entirely sure but a bit concerned is that Flink may take some important actions in `CheckpointCoordinator#restoreLatestCheckpointedStateInternal()` even if `verticesToRestart` if empty, e.g. `invoking OperatorCoordinator#resetToCheckpoint(...)`. These actions were always taken previously, while are possible to be skipped after this change(when a global failover and regional failover happen concurrently).

I haven't had the chance to examine it all over yet. It's appreciated if you can also help to examine this.

Reply:
There are three functions will be skipped after this change:
1. OperatorCoordinatorHolder#resetToCheckpoint
2. CheckpointCoordinator#abortPendingCheckpoints
3. MasterHooks#reset/restore

But those will be invoked later(new global failure) with the newest checkpoint-id. So i think this change is safe.

Reply:
A global failover can be superseded by a regional failover, regarding the tasks to restart.
Here's an example: Here's a job consists of one only pipelined region. A global failure happens first(caused by the OperatorCoordinator) and need to restart all the tasks. It also needs `OperatorCoordinatorHolder#resetToCheckpoint()` to be invoked to recover from an inconsistent status. However, a task happens later but almost at the same time, which needs to restart all the tasks. Therefore, the `verticesToRestart` would be empty when `restartTasks(...)` is invoked for the global failure. And `OperatorCoordinatorHolder#resetToCheckpoint()` will not be invoked.

Reply:
The global failover will change execution state to 'CANCELING' in #restartTasksWithDelay. So that the pending execution state transition(to 'FAILED') will not be executed and will not trigger a region failover. 

Therefore, only regional failover or global failover can be superseded by global failover.

Reply:
Yes you are right. If a global failover is in progress, no regional failover will be triggered.
Thanks for the explanation!